src/Services/Dominion/Actions/Military/TrainActionService.php

```diff
@@ -187,4 +168,52 @@ public function train(Dominion $dominion, array $data): array
             ],
         ];
     }
+    
+    /**
+     * Returns training message for a train action.
+     *
+     * @param Dominion $dominion
+     * @param array $unitsToTrain
+     * @param array $totalCosts
+     * @return string
+     */
+    private function getTrainingMessage(Dominion $dominion, $totalCosts) {
+        $unitsToTrainStringParts = [];
+
+       foreach ($unitsToTrain as $unitType => $amount) {
+            if($amount > 0) {
+                $race = strtolower($this->unitHelper->getUnitName($unitType, $dominion->race));
+                $unitsToTrainStringParts[] = number_format($amount) . ' ' . str_plural($race, $amount);
+            }
+       }
+        
+       $unitsToTrainString = $this->generateSentenceFromArray($unitsToTrainStringParts);
+
+       $trainingCostsStringParts = [];
+       foreach($totalCosts as $costType => $cost) {
+           if($cost > 0){
+             $trainingCostsStringParts[] = number_format($cost) . ' ' . str_plural($costType, $cost);
```

Comment:
The cost types 'platinum' and 'ore' get str_plural()'d here to 'platinums' and 'ores', respectively. Could you verify and change it it doesn't plural those terms?

The terms 'platinum' and 'ore' and always need to be singular for consistency reasons. Just feel free to throw a check like `in_array($costType, ['platinum', 'ore'], true)` in there.

Also if possible, please continue/break/return early to prevent if-nesting.

```PHP
if ($cost === 0) {
    continue;
}

$trainingCostsStringParts[] = ...
```